azure-iot-device/azure/iot/device/common/transport/mqtt/pipeline_stages.py

```diff
@@ -0,0 +1,645 @@
+
+        def single_op_callback(completed_op):
+            logger.info(
+                "serial: %s(%s) complete. %s items left",
+                self.name,
+                completed_op.name,
+                len(worklist),
+            )
+            if completed_op.error:
+                last_op_done(completed_op)
+            else:
+                if len(worklist) == 0:
+                    last_op_done(completed_op)
+                else:
+                    next_op = worklist.popleft()
+                    next_op.callback = single_op_callback
+                    self.run_op(next_op)
+
+        first_op = worklist.popleft()
+        first_op.callback = single_op_callback
+        self.run_op(first_op)
+
+    @abc.abstractmethod
+    def _run_op(self, op):
+        """
+        Abstract method to run the actual operation.  Implemented in derived classes.
+        """
+        pass
+
+    def handle_pipeline_event(self, event):
+        """
+        Handle a pipeline event that arrives from the stage below this stage.  This
+        is the function that is called from outside this object.  Stages which are
+        derived from this Class should implement specific event handling code in the
+        private _handle_pipeline_event function.
+        """
+        try:
+            self._handle_pipeline_event(event)
+        except Exception as e:
+            logger.error(msg="Error in _handle_pipeline_event call", exc_info=e)
+            self.pipeline_root.unhandled_error_handler(e)
+
+    def _handle_pipeline_event(self, event):
+        """
+        Function which actually handles an event.  If a stage wants to handle events,
+        it should override this function.
+        """
+        if self.previous:
+            self.previous._handle_pipeline_event(event)
+
+    def continue_op(self, op):
+        """
+        Helper function to continue a given operation by passing it to the next stage
+        in the pipeline.  If there is no next stage in the pipeline, this function
+        will fail the operation and call complete_op to return the failure back up the
+        pipeline.  If the operation is already in an error state, this function will
+        complete the operation in order to return that error to the caller.
+        """
+        if op.error:
+            self.complete_op(op)
+        elif not self.next:
+            logger.error("%s(%s): no next stage.  Returning error", self.name, op.name)
+            op.error = NotImplementedError()
+            self.complete_op(op)
+        else:
+            self.next.run_op(op)
+
+    def complete_op(self, op):
+        """
+        Helper function to complete an operation by calling it's callback function thus
+        returning the result of the operation back up the pipeline.
+        """
+        logger.info("{}({}): completing".format(self.name, op.name))
+        op.callback(op)
+
+    def continue_with_translated_op(self, original_op, translated_op):
```

Comment:
>continue_with_translated_op [](start = 8, length = 27)

Translated Operation not defined as a term here - I have no idea what this does from the docs.

Additionally, while I can infer what it does based on my knowledge, this concept of application specific term (enable C2D) vs. mqtt specific operation (subscribe), is elsewhere referred to as "conversion", which is a confusing mix of terms (if they are the same thing)

Reply:
I've changed names and improved the doc.  I'm not sure if I'm ready to introduce a new subclass yet.  Conversion is the most likely reason to use this function, but I think we'll find other uses.

---
In reply to: [279661803](https://github.com/Azure/azure-iot-sdk-python-preview/pull/87#discussion_r279661803) [](ancestors = 279661803)